lib/sequence.js

```diff
@@ -250,20 +274,29 @@ Sequence.prototype._resetCounter = function(id, reference, callback){
  * @param      {Function}  callback  Called with the sequence counter
  */
 Sequence.prototype._setNextCounter = function(doc, callback) {
-
   var retriable = function(callback) {
     var id = this.getId();
     var referenceValue = this._getCounterReferenceField(doc);
-    this._counterModel.findOneAndUpdate(
-            { id: id, reference_value: referenceValue },
-            { $inc: { seq: 1 } },
-            { new: true, upsert: true },
-            function(err, counter) {
-              if (err) return callback(err);
-              return callback(null, counter.seq);
-            }
-
+    var incrementAmount = this.getIncrementAmount();
+
+    this._counterModel.countDocuments({id: id, reference_value: referenceValue}).then((count) => {
+      if (count > 0 || (incrementAmount === 1 && this._options.start_seq === 1)) {
```

Comment:
Tested this out and doesn't seem to work. Looks like if we use the same fields in update and insert setDefaultsOnInsert doesn't work. More details here: https://github.com/Automattic/mongoose/issues/3617

Reply:
An alternative approach would be to initialise the counter model with all default values during startup so that we can safely execute increments queries only.

Reply:
This sounds as a good idea. The only drawback is that, when after the counter collection is instantiated, the starting value cannot be changed anymore. Not a big issue, just something to document. 
What would be the initial value, `start_seq - 1`?

Reply:
I think that drawback is something that we could live with for now.
Shouldn't initial value be start_seq - inc_amount ?

Reply:
> Shouldn't initial value be start_seq - inc_amount ?

Yes, indeed

> I just realised that we cant initialise counters for scoped increments where the reference_fields are only known during runtime

The scoped increments are more a pain than a feature. I know about people using it but it prevents me to implement other features or fix some bug because any fix wasn't applicable to scoped counters. I'm already thinking to create a new major version and removing them but I don't know when this will happen.

Maybe the simplest for the moment would be to forbid custom `start_seq` for scoped counters, while the custom `inc_amount` would probably work anyway right?

Side question: just to understand, are you @Pr0Sh00t3r related to @tscolbe?

Reply:
No, we are not related, was searching for this feature and I found this PR.
The scoped increments are why I am using this plugin in the first place, please don't remove it.

And I think i fixed the start_seq even for scoped increments. So the idea is to initialise the counter entry before increment, but not on startup as i suggested before, instead right before the _setNextCounter hook. This may affect performance slightly as there is one more DB call in hook but it actually works.

something like this:
```
sequence._createCounter(this, (err, seq) => {
                if (err) {
                    callback(err);
                    return;
                }
                if (!_.isNull(seq)) {
                    this[sequence._options.inc_field] = seq;
                    cb();
                } else {
                    sequence._setNextCounter(this, (err, seq) => {
                        if (err) {
                            cb(err);
                            return;
                        }
                        this[sequence._options.inc_field] = seq;
                        cb();
                    });
                }
            });
```
_createCounter would return the seq number if created, else null and _setNextCounter would work as usual.

@ramiel Do you see any issues with this approach ?

Reply:
I'm not really sure to understand what you mean. When is this `_createCounter` called? If we call when we want to increment, it means we should call it inside `_setNextCounter` which is called by this in turn... Anyway, the general idea of setting the counter collection on the first increment is good but are we breaking the rule of atomic updates?

Reply:
Well, we could call it inside _setNextCounter itself. Just wrote it as a separate schema function for the sake of clarity.

```
   Sequence.prototype._createCounter = function(doc, callback) {
        const id = this.getId();
        const reference_value = this._getCounterReferenceField(doc);
        const startSeq = this._options.start_seq;
        const counterModel = this._counterModel;

        counterModel.findOneAndUpdate(
            {
                id,
                reference_value
            },
            {},
            {
                upsert: true,
                new: true,
                setDefaultsOnInsert: true
            },
            (err, counter) => {
                if (err) return callback(err);
                // upserted is present if entry was created now
                if (_.has(counter, "upserted") && counter.upserted.length > 0)
                    return callback(null, startSeq);
                else return callback(null, null);
            }
        );
    };
```

This is how I wrote it and works. I only used findOneAndUpdate so I believe it is atomic.

Reply:
Yes, the operatios itself is atomic, but not the operation of setting the next counter. Anyway, would you like to propose a PR (or update this one)

Reply:
Will submit a new PR